### DISCUSSION TITLE
List records with nested filters on relation fields

### DISCUSSION THREAD

**Discussion by shubhendumadhukar** [pocketbase/pocketbase#3335] [Q&A]
I have an app built on pocketbase, which has 2 collection, `chats` and the default collection `users`. I am trying to search specific chats based on some filters, but I am running into some weird issues. Looking for some help with it.

### My Setup
My search filter is `title ~ "${query}" || participants.name ?~ "${query}" || participants.username ?~ "${query}" || participants.email ?~ "${query}"`, where query is the search string.

`participants` is a multiple relation field on chats which points to the users collection.

On users collection list rules is empty i.e. anyone can list the users. Whereas on chats, list rule is `@request.auth.id ?= participants.id`

### Problem
Let's say I have 1 chat in which there are two participants, userA and userB. I am logged in with userA.

1. When I search chats without any filters, I get the expected list of chats.
2. When I search chats with query set as userA's name (the logged-in user), I get the expected list of chats.
3. When I search chats with query set as userB's name, **I get an empty array.**

I am sure it has something to do with the rules specified on the collections, since if I remove the list rule specified on the chats collection, I get expected results. But I need that rule so that users only see the chats they are participating in. I can't figure out what else to change. TIA for any help.

**Comment by ganigeorgiev** [comment]
The result in step3 comes from the fact that internally the API rule and client-side filter are resolved in a single run for optimization purposes.

`@request.auth.id ?= participants.id` will join **only** the participants records that matches the constraint.
The search filter is applied **in addition** to the above, meaning that it will work only on the participants records that were already previously joined.

To some extend this can be avoided if we had support for exact `has/in` array value match (_it exists somewhere in my todo and I'll reconsider prioritizing it for the next v0.19/v0.20_), but depending on your use case one possible workaround at the moment could be to change your list rule to `participants ~ @request.auth.id` to perform a loose like/contains search on the raw json array value.

I haven't tried it and I'm not sure how well it'll perform, but another a little hacky workaround could be to manually join the chats collection with itself using the `@collection.*` filter in your rule so that the API rule and search filter joins are kept separate, something like:
```js
@collection.chats.id ?= id &&
@collection.chats.participants.id ?= @request.auth.id
```

**Reply by shubhendumadhukar** [reply]
Thank you. While your recommended approach (`participants ~ @request.auth.id`) solves my issue, I am not sure I understand why it works. The list, without any filters, looks the same for `participants ~ @request.auth.id` and `@request.auth.id ?= participants.id`, at least on the client side. It contains "chats" records, with the details of other participants too, not just the logged-in user. So I am a bit confused why the additional filters make the final list an empty array.

**Reply by ganigeorgiev** [reply]
When accessing a nested relation field (eg. `participants.id`) we perform a join with the related collection. `@request.auth.id ?= participants.id` and `participants.name ?~ "${query}"` will target the same joined records because the API rules and filters are applied together and joins are reused whenever possible in attempt to optimize the query performance.

`participants ~ @request.auth.id` on the other hand doesn't perform a join and doesn't interfere with the search filter. It is directly translated to something like `participants LIKE "%{:authId}%"`, where participants is the raw serialized json array of rel ids (eg. `["id1","id2","id3"]`).

**Reply by shubhendumadhukar** [reply]
Thank you for the explanation. That helps.

**Reply by ganigeorgiev** [reply]
I agree that it is not very intuitive and may have some unexpected side-effects similar to the above, especially considering that for the end users the 2 constraints (API rule and client-side filters) are defined in two different places.
I've added task in the roadmap to reevaluate the joins reuse in the near future (_I'll perform some benchmarks and eventually I'm planning to allow specifying join prefix(es) to separate the nested constraint checks_).
